robot_smach_states/src/robot_smach_states/util/designators/core.py

```diff
@@ -55,20 +71,32 @@ def resolve(self):
         result_type = type(result)
         resolve_type = self.resolve_type
 
-        if isinstance(result, list):
-            if not result:
-                return []
-            result_type = type(result[0])
+        if result is None: # No data available for checking the type.
+            return result
+
+        if isinstance(self._resolve_type, list):
+            # Not None, list expected instead.
+            if not isinstance(result, list):
+                self.fail_with_type_error(result_type, resolve_type, result)
+
+            if not result: # Empty list, cannot check element type.
+                return result
+
+            if not isinstance(result[0], resolve_type[0]): # Element type check.
+                self.fail_with_type_error(result_type, resolve_type, result)
 
-        if isinstance(self.resolve_type, list):
-            resolve_type = self.resolve_type[0]
+            return result
+
+        # Normal type.
+        if not issubclass(result_type, resolve_type):
+            self.fail_with_type_error(result_type, resolve_type, result)
 
-        if result is not None and not issubclass(result_type, resolve_type):
-            raise TypeError("{} resolved to a '{}' instead of expected '{}'. "
-                            "Originals: result type: {}, resolve type: {}".format(self, result_type, resolve_type,
-                                                                                  type(result), self.resolve_type))
         return result
 
+    def fail_with_type_error(self, result_type, resolve_type, result):
+            msg = "{} resolved to a '{}' instead of expected '{}'. Originals: result type: {}, resolve type: {}"
+            raise TypeError(msg.format(self, result_type, resolve_type, type(result), self.resolve_type))
```

Comment:
resolve_type = self.resolve_type = property(_get_resolve_type) = return self._resolve_type

I think resolve type is now printed two times, both time provided via a completely different route. This should be simplified.

Both calls to `fail_with_type_error` have the same arguments, so t

Reply:
Some text seems to be missing at the end?

I moved some `[0]` index operations it seems, which makes it print duplicate information more often than it did in the past. I'll look into it.